_posts/2016-05-01-Using-JWT-Tokens-for-State-Transfer.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: post
+author: andreas_evers
+title: 'Using JWT tokens for state transfer'
+image: /img/JWT/jwt-logo.png
+tags: [JWT,State,Cache]
+category: Microservices
+comments: true
+---
+
+At one of our clients, we have been using JWT tokens quite extensibly.
+We even use it to persist state on the client.
+
+## Why persisting state on the client?
+
+When buiding microservices, we need to build so-called "cloud native" applications.
+One of the key tenents of cloud native application design is keeping your services stateless.
+The benefit of having stateless applications is foremost the ability to respond to events by adding or removing instances without needing to significantly reconfigure or change the application.
+More stateless services can easily be added when load suddenly increases, or if an existing stateless service fails, it can simply be replaced with another.
+Hence, resilience and agility, are easier to achieve with stateless services.
+
+Keeping your services stateless means we need to persist our state somewhere else.
+Since we are tranferring state in the REST architectural style, we can use the client to retain our state.
+For scaling purposes this is a great solution, as the client will only ever have to store its own state, and the server will be relieved of the state of all its clients.
+
+At our client we have chosen to use JWT tokens for this state transfer to the client.
+While JWT is primarily intended for authentication and authorization purposes, the specification allows us to add any data we'd like to verify later on.
+
+## Looking good
+
+Imagine the following scenario:
+
+A list of products is fetched from the products microservice.
+The user isn't allowed to view all products, so only those products the user has access to are returned.
+When the user wants to order a product, he sends an order request to the order microservice with the id of the product he wants to order.
+At that moment the orders microservice needs to know whether or not the user is allowed to order access this product, let alone order it.
```

Comment:
+At that moment the "orders microservice"

And so on for all "identifying" applications.